File: 2018/csharp/day08.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace aoc2018
{
    class Day8
    {
        public static void Solve()
        {
            Queue<int> Q = new Queue<int>();
            File.ReadAllText(Path.Combine("..", "input", "day8.input")).Split(" ").Select(int.Parse).ToList().ForEach(x => Q.Enqueue(x));
            int sum = 0;
            Node root = BuildNode(Q, ref sum);
            Console.WriteLine("day8, part a: {0}", sum);
            Console.WriteLine("day8, part b: {0}", GetNodeValue(root));
        }

        public static int GetNodeValue(Node node)
        {
            if (node.children.Count > 0)
            {
                int value = 0;
                node.values.Where(x => x > 0 && x <= node.children.Count).ToList().ForEach(x => value += GetNodeValue(node.children[x - 1]));
                return value;
            }
            else
                return node.values.Sum();
        }

        public static Node BuildNode(Queue<int> Q, ref int sum)
        {
            int countChildren = Q.Dequeue(), countValues = Q.Dequeue();
            var node = new Node { children = new List<Node>(), values = new List<int>() };
            for (int i = 0; i < countChildren; i++)
                node.children.Add(BuildNode(Q, ref sum));
            for (int i = 0; i < countValues; i++)
            {
                node.values.Add(Q.Dequeue());
                sum += node.values[node.values.Count - 1];
            }
            return node;
        }
    }

    struct Node
    {
        public List<Node> children;
        public List<int> values;
    }
}

File: 2018/python/day08.py
def next_val():
    global ix, input
    ix += 1
    return input[ix -1]

def buildNode():   
    children, metadata = [], []
    cc, cm = next_val(), next_val() 
    for _ in range(cc):
        children.append(buildNode())
    for _ in range(cm):
        metadata.append(next_val())
    return children, metadata

def sumMeta(C, M):
    return sum(sumMeta(c[0], c[1]) for c in C) + sum(M)

def sumValue(C, M):
    return sum(M) if len(C) == 0 else sum(sumValue(C[m-1][0], C[m-1][1]) for m in M if m > 0 and m <= len(C))

input = list(map(int, open("../input/day8.input").readline().strip("\n").split(" ")))
ix = 0
root = buildNode()
print("day 8, part a:", sumMeta(root[0], root[1]))
print("day 8, part b:", sumValue(root[0], root[1]))